server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -1406,7 +1412,7 @@ public Node start() throws NodeValidationException {
             injector.getInstance(PersistedClusterStateService.class),
             pluginsService.filterPlugins(ClusterCoordinationPlugin.class)
         );
-        if (Assertions.ENABLED) {
+        if (Assertions.ENABLED && DiscoveryNode.isStateless(settings()) == false) {
```

Comment:
This is a bit unrelated, I can open a separate PR if you think it's best.

Reply:
I think we should find a better way in due course, could you open an issue and add a TODO here?

Reply:
I was wondering if the legacy format still applies in 8.x versions?

Reply:
There is some BwC cruft in this area for sure, but we'd need to spend some time looking at the details (and the timeline of changes) to come up with a plan for cleaning it up.

Reply:
Addressed in https://github.com/elastic/elasticsearch/pull/95206/commits/20baea563083d2d3e67be83d2871cab05f99183e